registry/cache.go

```diff
@@ -226,18 +302,24 @@ func (rc *registryCache) PushManifest(ctx context.Context, repo string, tag stri
 }
 
 func (rc *registryCache) PushBlob(ctx context.Context, repo string, desc ociregistry.Descriptor, r io.Reader) (ociregistry.Descriptor, error) {
-	// TODO this does *not* need to lock the entire cache during the upstream push (but it *would* be good to block pushing to this specific digest)
-	rc.mu.Lock()
-	defer rc.mu.Unlock()
+	digest := desc.Digest
+	digestKey := cacheKeyDigest(repo, digest)
 
-	// TODO if desc.Size <= manifestSizeLimit, we should technically wrap up the Reader we're given and cache the result so we can shove it directly into the cache, but we currently don't read back blobs we pushed in (and I don't think that's a common use case), so I'm taking the simpler answer of just using this event as a cache bust intead
+	refMu := rc.refMutex(digestKey)
+	refMu.Lock()
+	defer refMu.Unlock()
```

Comment:
These usages are very similar. Is it possible to refactor to reduce duplication?

Reply:
The best I could do is reduce three lines to two (by having `refMutex` immediately invoke `.Lock()`), which IMO makes it less obvious that we've used the mutex correctly (because you would then have to trace into the function call to see the call to `Lock` to verify that every `Lock` has a matching `Unlock`).